tests/functional/test_csv_viewer.py

```diff
@@ -0,0 +1,108 @@
+    # test run and occasionally caching causes tests to fail. So instead
+    # we create a new context and page for each test run
+    context = browser.new_context()
+    page = context.new_page()
+    workspace = factories.create_workspace("my-workspace")
+
+    factories.write_workspace_file(
+        workspace,
+        "outputs/file1.csv",
+        csv_file,
+    )
+    login_as_user(
+        live_server,
+        context,
+        user_dict=factories.create_api_user(
+            username="author",
+            workspaces={
+                "my-workspace": factories.create_api_workspace(project="Project 2"),
+            },
+        ),
+    )
+
+    page.goto(
+        live_server.url + workspace.get_contents_url(UrlPath("outputs/file1.csv"))
+    )
+
+    reader = csv.reader(StringIO(csv_file), delimiter=",")
+    csv_list = []
+    for row in reader:  # each row is a list
+        csv_list.append(row)
+
+    failures = page.evaluate(
+        """(csv_list) => {
+            let failures = [];
+
+            // First we get the contents of the header row
+            const headerRow = document.querySelector('thead tr');
+            const headerRowCells = Array.from(headerRow.querySelectorAll('th'));
+            // This assumes the first column are the row numbers so we ignore them
+            const headerValues = headerRowCells.map(cell => cell.textContent.trim()).slice(1);
+
+            // Next we get the table rows
+            const bodyRows = Array.from(document.querySelector('tbody').querySelectorAll('tr'));
+            const rowValues = bodyRows.map((row) => {
+                return Array.from(row.querySelectorAll('td')).map(cell => cell.textContent).slice(1)
+            });
+
+            // We add the headers back to the rowValues array as the first row
+            rowValues.unshift(headerValues);
+
+            // For each row in the actual csv file, we check that it is the same
+            // as what is displayed in the table.
+            csv_list.forEach((row, i) => {
+              if(JSON.stringify(row) !== JSON.stringify(rowValues[i])) {
+                failures.push(`The csv row ${JSON.stringify(row)} does not match the table row ${JSON.stringify(rowValues[i])}. In case it's useful the table row is ${rowValues[i]} before passed to JSON.strinfigy.`)
+              }
+            })
+
+            // Finally we check that the number of rows in the csv file matches the html table
+            if(csv_list.length !== rowValues.length) {
+              failures.push(`CSV has ${csv.length} rows, but HTML table has ${rowValues.length}`);
+            }
+            return failures;
+          }
+        """,
+        csv_list,
+    )
+
+    assert len(failures) == 0, print(repr(failures))
+
+    # In theory we should tidy up by closing the context. But there is a bug which
+    # causes this test to hang when run with coverage turned on. So we check to see
+    # if coverage is on, and if not we tidy up.
```

Comment:
o_O

Reply:
Is there an issue on the playwright or coverage repo we can link to?

Reply:
I had found one for hypothesis (https://github.com/HypothesisWorks/hypothesis/issues/4052) which I'll link to.

Reply:
Oh,looks like we might need to upgrade our Python version.